src/vsr/replica.zig

```diff
@@ -1270,14 +1293,11 @@ pub fn ReplicaType(
 
         fn on_start_view_change(self: *Self, message: *Message) void {
             assert(message.header.command == .start_view_change);
+            self.view_jump(message.header);
             if (self.ignore_start_view_change_message(message)) return;
```

Comment:
> We probably should change if (header.view < self.view) return; to assert and just consistently ignore before jump (I see that we don't have ignore_commit / ignore_prepare yet, but we probably should)

:+1: That works. However,

> It makes sense to try to ignore a message before reacting to it.

In the case of a `view_jump`, sometimes we want to `view_jump` even when we ignore. That is, we often want to ignore the message from a newer view, but still increase our own view (or request a SV).

This is one of those cases:
https://github.com/tigerbeetledb/tigerbeetle/pull/561/commits/d32b3edf04b648b0c654e375f687942ba7c947ac

Reply:
> In the case of a view_jump, sometimes we want to view_jump even when we ignore. 

Hm, maybe then we should always view_jump before checking if the message should be ignored? Although I don't like it -- view_jumping on, eg, misdirected messages should be correct, but I'd rather we not do that, to simplify the state space.

Reply:
I agree that we should be consistent about when we `view_jump`.

What about:
- Split `view_jump` into `view_jump_to_normal` (the RSV case) and `view_jump_to_view_change` (the transition-to-view-change case).
- `view_jump_*()` returns `bool` — whether or not `message.header.view != self.view`. If `true`, `return` from the message handler (it is like an extra `ignore_` helper, which as a side effect will try to RSV/transition).
- Change `ignore_*` to assert that `view` matches current view.

That likewise would jump on misdirected messages too, though. :thinking: 
I guess we could filter out `message.header.replica == replica` easily enough in `view_jump_*`.

Reply:
Or `ignore_{message}`/`on_{message}` could `error { Misdirected, ViewOld, ViewNew }!void`, and `on_message` would turn `ViewNew` into a `view_jump`.

Reply:
Another option is to kick `view_jump` from within ignore_message. This makes sense from control-flow perspective, but is not great given the current names.

Reply:
Also I like how the current `ignore_` functions have no side effects

Reply:
I don't think we should split `view_jump()` across sub-protocols, but would rather that we keep it all contained.

Also, would not like to increase the dimensionality of the `ignore_()` functions by moving from boolean to error set. The boolean return type is less dimensionality at the call site.

Let's keep this as is.

But it would be good to document why we always check to see if we should view jump, but may then decide thereafter to ignore a message.

Reply:
"less dimensionality at the call site" is the primary the reason I think that `view_jump` should be split.
The caller often needs to know which branch is being taken, but previously that was only implicitly known (based on the context, i.e. the message command).

The other reason is that the two branches don't actually share code.

See:

```zig
        fn view_jump_to_normal(self: *Self, header: *const Header) void {
            assert(header.command == .prepare or header.command == .commit or
                header.command == .start_view_change);

            if (header.command == .start_view_change) {
                // When we are recovering_head we can't participate in a view-change anyway.
                // But there is a chance that the primary is still running and reachable,
                // despite the SVC.
                assert(self.status == .recovering_head);
            }

            if (header.view < self.view) return;

            // We need a start_view from any other replica — don't request it from ourselves.
            if (self.primary_index(header.view) == self.replica) return;

            switch (self.status) {
                // If the transition is to `.normal`, then ignore if for the same view:
                .normal => if (header.view == self.view) return,
                // This is an interesting special case:
                // If the transition is to `.normal` in the same view, then we missed the
                // `start_view` message and we must also consider this a view jump:
                .view_change => {},
                .recovering_head => {},
                .recovering => return,
            }

            if (header.view == self.view) {
                assert(self.status == .view_change or self.status == .recovering_head);

                log.debug("{}: view_jump: waiting to exit view change", .{self.replica});
            } else {
                assert(header.view > self.view);
                assert(self.status == .view_change or self.status == .recovering_head or
                    self.status == .normal);

                log.debug("{}: view_jump: waiting to jump to newer view", .{self.replica});
            }

            // TODO Debounce and decouple this from `on_message()` by moving into `tick()`:
            // (Using request_start_view_message_timeout).
            log.debug("{}: view_jump: requesting start_view message", .{self.replica});
            self.send_header_to_replica(self.primary_index(header.view), .{
                .command = .request_start_view,
                .cluster = self.cluster,
                .replica = self.replica,
                .view = header.view,
            });
        }

        fn view_jump_to_view_change(self: *Self, header: *const Header) void {
            assert(self.status == .normal or self.status == .view_change);
            // status=recovering_head cannot transition to view-change — without a head, it cannot
            // safely persist its log suffix headers.
            assert(self.status != .recovering_head);
            assert(header.command == .do_view_change or header.command == .start_view_change);

            if (header.view <= self.view) return;

            if (header.view == self.view + 1) {
                log.debug("{}: view_jump: jumping to view change", .{self.replica});
            } else {
                log.debug("{}: view_jump: jumping to next view change", .{self.replica});
            }
            self.transition_to_view_change_status(header.view);
        }
```

Aside from the conceptual similarity of "response to higher view number" they don't actually have anything in common.

Reply:
Thanks @sentientwaffle,

By "reducing dimensionality at the call site" I was referring to increasing "the dimensionality of the ignore_() functions by moving from boolean to error set."

i.e. `error { Misdirected, ViewOld, ViewNew }!void` is more dimensionality than `void` or `boolean` so I don't prefer it to what we already have at present. We used to have error set return types and/or at least optional return types in an early version of VSR and I actively got rid of them to go for only `void` (best) or `boolean` return types, to reduce branching at the call site.

This experience was the inspiration for this bit in TigerStyle:

> Use simpler function signatures and return types to reduce dimensionality at the call site, the number of branches that need to be handled at the call site, because this dimensionality can also be viral, propagating through the call chain. For example, as a return type, void trumps bool, bool trumps u64, u64 trumps ?u64, and ?u64 trumps !u64.

Regarding `view_jump()`, the rationale is that:

* It's not meant as an imperative (e.g. jump to view change or jump to normal), but rather as a listener to jump if necessary, but in a way that the caller makes its own decisions thereafter.
* The caller shouldn't need to decide which version of the function to call.
* I also wanted to keep all the view jump logic contained and together, so that one can reason about all the enumerations (given source/target sub-protocols) in one place. Spreading these out increases the risk of a spatial TOCTOU.

Reply:
> It's not meant as an imperative (e.g. jump to view change or jump to normal), but rather as a hook/listener to jump if necessary, but in a way that the caller makes its own decisions thereafter, without relying on any return value. This is the least coupling.

Gotcha, that makes sense.

In that case though, should `view_jump` move up into `on_message` so that each message handler does not need to call it separately? And then rather than asserting `header.command`, `view_jump` would just ignore irrelevent commands (i.e. any command without an attached `view`).

Reply:
Great idea, let's do it.